### PR TITLE
feat: find the highest common API version FS-451

### PR DESCRIPTION
**Pull request by sb88k**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-451" title="FS-451" target="_blank">Sub-taskFS-451</a>  [iOS] Use the common version for all API calls
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

Apply changes in ZMAPIVersion from Transport.
Add function to find the highest common version between the client and backend supporting versions.

### CODE REVIEW

File: Sources/Helpers/APIVersion.swift [hunk 1]
```diff
@@ -48,12 +48,14 @@ extension ZMAPIVersion {
...
 
-extension ZMAPIVersion: Comparable {
+extension APIVersion {
 
-    public static func < (lhs: Self, rhs: Self) -> Bool {
-        return lhs.rawValue < rhs.rawValue
+    public static func highestCommonVersion(with versions: [Int32]) -> Self? {
+        versions.sorted(by: >)
+            .compactMap { APIVersion(rawValue: Int32($0)) }
+            .first
```
Comment (johnxnguyen) [inline Sources/Helpers/APIVersion.swift:58]: An alternate way to express this, maybe slightly more readable?

```suggestion
        versions
            .compactMap { APIVersion(rawValue: Int32($0)) }
            .max()
```

File: Sources/Helpers/APIVersionTests.swift [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        highestVersion = nil
+    }
+
+    func testThatTheCommonVersionIsNil_whenBackendSupportsHigherVersionsOnly() {
+        // given
+        var backendVersions = [Int32]()
+
+        for version in highestVersion.rawValue + 1 ..< highestVersion.rawValue + 5 {
+            backendVersions.append(version)
+        }
```
Comment (johnxnguyen) [inline Sources/Helpers/APIVersionTests.swift:41]: I think you can simplify this with:

```suggestion
        var backendVersions = (1...5).map { higestVersion + $0 }
```

File: Sources/Helpers/APIVersionTests.swift [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        XCTAssertNil(commonVersion)
+    }
+
+    func testThatTheCommonVersionIsTheHighstVersion_whenBackendSupportsTheHighestVersion() {
+        // given
+        var backendVersions = [Int32]()
+
+        for version in lowestVersion.rawValue ... highestVersion.rawValue {
+            backendVersions.append(version)
+        }
```
Comment (johnxnguyen) [inline Sources/Helpers/APIVersionTests.swift:71]: You can also simplify this:

```suggestion
        var backendVersions = APIVersion.allCase.map(\.rawValue)
```

File: Sources/Helpers/APIVersionTests.swift [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        XCTAssertNil(commonVersion)
+    }
+
+    func testThatTheCommonVersionIsNil_whenBackendSupportsLowerVersionsOnly() {
+        // given
+        var backendVersions = [Int32]()
+
+        for version in highestVersion.rawValue - 5 ..< highestVersion.rawValue - 1 {
+            backendVersions.append(version)
+        }
```
Comment (johnxnguyen) [inline Sources/Helpers/APIVersionTests.swift:52]: You can also simply this one too:

```suggestion
        var backendVersions = ((-5)...(-1)).map { highestVersion.rawValue + $0 }
```

**Review by johnxnguyen — APPROVED** [review]

File: Sources/Helpers/APIVersion.swift [hunk 5]
```diff
@@ -48,12 +48,14 @@ extension ZMAPIVersion {
...
 
-// MARK: - Comparable
+// MARK: - Find the highest common version
 
-extension ZMAPIVersion: Comparable {
+extension APIVersion {
 
-    public static func < (lhs: Self, rhs: Self) -> Bool {
-        return lhs.rawValue < rhs.rawValue
+    public static func highestCommonVersion(with versions: [Int32]) -> Self? {
```
[the remark this answers is not in the document]
Comment (johnxnguyen) [reply Sources/Helpers/APIVersion.swift:55]: Or perhaps `ZMAPIVersion.highestSupportedVersion(in: versions)`

**Review by johnxnguyen — APPROVED** [review]

File: Cartfile.resolved [hunk 6]
```diff
@@ -12,5 +12,5 @@ github "wireapp/wire-ios-link-preview" "34.1.0"
 github "wireapp/wire-ios-protos" "30.1.1"
 github "wireapp/wire-ios-system" "36.1.1"
 github "wireapp/wire-ios-testing" "27.1.3"
-github "wireapp/wire-ios-transport" "feature/api-version"
+github "wireapp/wire-ios-transport" "feature/api-version-find-common"
```
Comment (johnxnguyen) [inline Cartfile.resolved:15]:
```suggestion
github "wireapp/wire-ios-transport" "feature/api-version"
```

**Review by David-Henner — APPROVED** [review]

**Pull request merged**